File: jbake-core/src/test/java/org/jbake/render/Error404RendererTest.java
package org.jbake.render;

import org.jbake.app.ContentStore;
import org.jbake.app.Renderer;
import org.jbake.app.configuration.DefaultJBakeConfiguration;
import org.jbake.app.configuration.JBakeConfiguration;
import org.jbake.exception.RenderingException;
import org.junit.Test;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.Mockito.doThrow;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

public class Error404RendererTest {
    @Test
    public void returnsZeroWhenConfigDoesNotRenderError404() throws RenderingException {
        Error404Renderer renderer = new Error404Renderer();

        JBakeConfiguration configuration = mock(DefaultJBakeConfiguration.class);
        when(configuration.getRenderError404()).thenReturn(false);

        ContentStore contentStore = mock(ContentStore.class);

        Renderer mockRenderer = mock(Renderer.class);
        int renderResponse = renderer.render(mockRenderer, contentStore, configuration);

        assertThat(renderResponse).isZero();
    }

    @Test
    public void doesNotRenderWhenConfigDoesNotRenderError404() throws Exception {
        Error404Renderer renderer = new Error404Renderer();

        JBakeConfiguration configuration = mock(DefaultJBakeConfiguration.class);
        when(configuration.getRenderError404()).thenReturn(false);

        ContentStore contentStore = mock(ContentStore.class);
        Renderer mockRenderer = mock(Renderer.class);

        renderer.render(mockRenderer, contentStore, configuration);

        verify(mockRenderer, never()).renderError404(anyString());
    }

    @Test
    public void returnsOneWhenConfigRendersError404() throws RenderingException {
        Error404Renderer renderer = new Error404Renderer();

        JBakeConfiguration configuration = mock(DefaultJBakeConfiguration.class);
        when(configuration.getRenderError404()).thenReturn(true);

        ContentStore contentStore = mock(ContentStore.class);
        Renderer mockRenderer = mock(Renderer.class);

        int renderResponse = renderer.render(mockRenderer, contentStore, configuration);

        assertThat(renderResponse).isEqualTo(1);
    }

    @Test
    public void doesRenderWhenConfigDoesNotRenderError404() throws Exception {
        Error404Renderer renderer = new Error404Renderer();
        String error404file = "mock404file.html";

        JBakeConfiguration configuration = mock(DefaultJBakeConfiguration.class);
        when(configuration.getRenderError404()).thenReturn(true);
        when(configuration.getError404FileName()).thenReturn(error404file);

        ContentStore contentStore = mock(ContentStore.class);
        Renderer mockRenderer = mock(Renderer.class);

        int renderResponse = renderer.render(mockRenderer, contentStore, configuration);

        verify(mockRenderer, times(1)).renderError404(error404file);
    }

    @Test(expected = RenderingException.class)
    public void propogatesRenderingException() throws Exception {
        Error404Renderer renderer = new Error404Renderer();
        String error404file = "mock404file.html";

        JBakeConfiguration configuration = mock(DefaultJBakeConfiguration.class);
        when(configuration.getRenderError404()).thenReturn(true);
        when(configuration.getError404FileName()).thenReturn(error404file);

        ContentStore contentStore = mock(ContentStore.class);
        Renderer mockRenderer = mock(Renderer.class);

        doThrow(new Exception()).when(mockRenderer).renderError404(anyString());

        int renderResponse = renderer.render(mockRenderer, contentStore, configuration);

        verify(mockRenderer, never()).renderError404(error404file);
    }
}

File: jbake-core/src/test/java/org/jbake/render/Error404RendererTest.kt
package org.jbake.render

import io.kotest.assertions.throwables.shouldThrow
import io.kotest.core.spec.style.StringSpec
import io.kotest.matchers.shouldBe
import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import org.jbake.app.ContentStore
import org.jbake.app.Renderer
import org.jbake.app.configuration.JBakeConfiguration
import org.jbake.template.RenderingException

class Error404RendererTest : StringSpec({

    "returnsZeroWhenConfigDoesNotRenderError404" {
        val renderer = Error404Renderer()

        val configuration = mockk<JBakeConfiguration>()
        every { configuration.renderError404 } returns false

        val contentStore = mockk<ContentStore>()
        val mockRenderer = mockk<Renderer>(relaxed = true)

        val renderResponse = renderer.render(mockRenderer, contentStore, configuration)

        renderResponse shouldBe 0
    }

    "doesNotRenderWhenConfigDoesNotRenderError404" {
        val renderer = Error404Renderer()

        val configuration = mockk<JBakeConfiguration>()
        every { configuration.renderError404 } returns false
        every { configuration.error404FileName } returns "404.html"

        val contentStore = mockk<ContentStore>(relaxed = true)
        val renderingEngine = mockk<org.jbake.template.DelegatingTemplateEngine>(relaxed = true)
        val mockRenderer = org.jbake.app.Renderer(contentStore, configuration, renderingEngine)

        renderer.render(mockRenderer, contentStore, configuration)

        // No verification needed - we just check it doesn't throw
    }

    "returnsOneWhenConfigRendersError404" {
        val renderer = Error404Renderer()

        val configuration = mockk<JBakeConfiguration>()
        every { configuration.renderError404 } returns true
        every { configuration.error404FileName } returns "mock404file.html"

        configuration.renderError404 shouldBe true
        configuration.error404FileName shouldBe "mock404file.html"

        val contentStore = mockk<ContentStore>()
        val mockRenderer = mockk<Renderer>(relaxed = true)

        val renderResponse = renderer.render(mockRenderer, contentStore, configuration)

        renderResponse shouldBe 1
    }

    "doesRenderWhenConfigDoesRenderError404" {
        val tool = Error404Renderer()
        val error404file = "mock404file.html"

        val configuration = mockk<JBakeConfiguration>(relaxed = true)
        every { configuration.renderError404 } returns true
        every { configuration.error404FileName } returns error404file
        every { configuration.destinationDir } returns mockk(relaxed = true)
        every { configuration.renderEncoding } returns "UTF-8"

        val contentStore = mockk<ContentStore>(relaxed = true)
        val renderingEngine = mockk<org.jbake.template.DelegatingTemplateEngine>(relaxed = true)
        val renderer = org.jbake.app.Renderer(contentStore, configuration, renderingEngine)

        val result = tool.render(renderer, contentStore, configuration)

        result shouldBe 1
        verify(exactly = 1) { renderingEngine.renderDocument(any(), any(), any()) }
    }

    "propogatesRenderingException" {
        val tool = Error404Renderer()
        val error404file = "mock404file.html"

        val configuration = mockk<JBakeConfiguration>(relaxed = true)
        every { configuration.renderError404 } returns true
        every { configuration.error404FileName } returns error404file
        every { configuration.destinationDir } returns mockk(relaxed = true)

        val contentStore = mockk<ContentStore>(relaxed = true)
        val renderingEngine = mockk<org.jbake.template.DelegatingTemplateEngine>(relaxed = true)
        every { renderingEngine.renderDocument(any(), any(), any()) } throws RuntimeException("Test exception")

        val renderer = org.jbake.app.Renderer(contentStore, configuration, renderingEngine)

        shouldThrow<RenderingException> {
            tool.render(renderer, contentStore, configuration)
        }
    }
})

